dnsmasq_interface.c

```diff
@@ -904,10 +863,85 @@ int FTL_listsfile(char* filename, unsigned int index, FILE *f, int cache_size, s
 		// Strip off everything at the end of the IP (CIDR might be there)
 		a=IPv6addr; for(;*a;a++) if(*a == '/') *a = 0;
 		// Prepare IPv6 address for records
-		if(inet_pton(AF_INET6, IPv6addr, &addr6) > 0)
-			has_IPv6 = true;
+		if(inet_pton(AF_INET6, IPv6addr, addr6) > 0)
+			*has_IPv6 = true;
 	}
 	clearSetupVarsArray(); // will free/invalidate IPv6addr
+}
+
+// Prototypes from functions in dnsmasq's source
+void add_hosts_entry(struct crec *cache, struct all_addr *addr, int addrlen, unsigned int index, struct crec **rhash, int hashsz);
+void rehash(int size);
+
+// This routine adds one domain to the resolver's cache. Depending on the configured blocking mode it may create
+// a single entry valid for IPv4 & IPv6 (containing only NXDOMAIN) or two entries one for IPv4 and one for IPv6
+// When IPv6 is not available on the machine, we do not add IPv6 cache entries (likewise for IPv4)
+static int add_blocked_domain_cache(struct all_addr *addr4, struct all_addr *addr6, bool *has_IPv4, bool *has_IPv6,
+                                    char *domain, struct crec **rhash, int hashsz, unsigned int index)
+{
+	int name_count = 0;
+	struct crec *cache4,*cache6;
+	// Add IPv4 record
+	if(has_IPv4 &&
+	   (cache4 = malloc(sizeof(struct crec) + strlen(domain)+1-SMALLDNAME)))
+	{
+		strcpy(cache4->name.sname, domain);
+		cache4->flags = F_HOSTS | F_IMMORTAL | F_FORWARD | F_REVERSE | F_IPV4;
+		// If we block in NXDOMAIN mode, we add the NXDOMAIN flag and make this host record
+		// also valid for AAAA requests
+		if(config.blockingmode == MODE_NX) cache4->flags |= F_IPV6 | F_NEG | F_NXDOMAIN;
+		cache4->ttd = daemon->local_ttl;
+		add_hosts_entry(cache4, addr4, INADDRSZ, index, rhash, hashsz);
+		name_count++;
+	}
+	// Add IPv6 record only if we respond with an IP address to blocked domains
+	if(has_IPv6 && config.blockingmode != MODE_NX &&
```

Comment:
Oh, you mean if their system doesn't even support IPv4 but only IPv6? I'm pretty sure that `dnsmasq` doesn't support this option at the moment (it will fail when it cannot bind to `0.0.0.0:53`). If they have no `IPV4_ADDRESS` in `setupVars.conf`, it will still work as we then use `0.0.0.0` as IPv4 address (see the first `if` block in `prepare_blocking_mode()`)